GVFS/GVFS.Common/NuGetUpgrade/NuGetUpgrader.cs

```diff
@@ -0,0 +1,578 @@
+                return false;
+            }
+
+            using (ITracer activity = this.tracer.StartActivity(nameof(this.TryDownloadNewestVersion), EventLevel.Informational))
+            {
+                try
+                {
+                    this.DownloadedPackagePath = this.nuGetFeed.DownloadPackageAsync(this.highestVersionAvailable.Identity).GetAwaiter().GetResult();
+                }
+                catch (Exception ex)
+                {
+                    this.TraceException(
+                        activity,
+                        ex,
+                        nameof(this.TryDownloadNewestVersion),
+                        "Exception encountered downloading newest version of upgrade package.");
+                    errorMessage = ex.Message;
+                    return false;
+                }
+            }
+
+            if (!this.noVerify)
+            {
+                if (!this.nuGetFeed.VerifyPackage(this.DownloadedPackagePath))
+                {
+                    errorMessage = "Package signature validation failed. Check the upgrade logs for more details.";
+                    this.tracer.RelatedError(errorMessage);
+                    this.fileSystem.DeleteFile(this.DownloadedPackagePath);
+                    return false;
+                }
+            }
+
+            errorMessage = null;
+            return true;
+        }
+
+        public override bool TryCleanup(out string error)
+        {
+            return this.TryRecursivelyDeleteInstallerDirectory(out error);
+        }
+
+        public override bool TryRunInstaller(InstallActionWrapper installActionWrapper, out string error)
+        {
+            string localError = null;
+            int installerExitCode;
+            bool installSuccessful = true;
+            using (ITracer activity = this.tracer.StartActivity(nameof(this.TryRunInstaller), EventLevel.Informational))
+            {
+                try
+                {
+                    string platformKey = InstallManifest.WindowsPlatformKey;
+
+                    if (!this.TryRecursivelyDeleteInstallerDirectory(out error))
+                    {
+                        return false;
+                    }
+
+                    if (!this.noVerify)
+                    {
+                        if (!this.nuGetFeed.VerifyPackage(this.DownloadedPackagePath))
+                        {
+                            error = "Package signature validation failed. Check the upgrade logs for more details.";
+                            activity.RelatedError(error);
+                            this.fileSystem.DeleteFile(this.DownloadedPackagePath);
+                            return false;
+                        }
+                    }
+
+                    this.UnzipPackage();
+                    this.installManifest = InstallManifest.FromJsonFile(Path.Combine(this.ExtractedInstallerPath, ContentDirectoryName, InstallManifestFileName));
```

Comment:
We could check if Unzip succeeded before trying to access extracted contents. (For a follow up PR)

Reply:
Based on my reading of the comments for ZipFile.ExtractToDirectory, it will throw if it encounters an error:

```
        // Exceptions:
        //   T:System.ArgumentException:
        //     destinationDirectoryName or sourceArchiveFileName is System.String.Empty, contains
        //     only white space, or contains at least one invalid character.
        //
        //   T:System.ArgumentNullException:
        //     destinationDirectoryName or sourceArchiveFileName is null.
        //
        //   T:System.IO.PathTooLongException:
        //     The specified path in destinationDirectoryName or sourceArchiveFileName exceeds
        //     the system-defined maximum length. For example, on Windows-based platforms, paths
        //     must not exceed 248 characters, and file names must not exceed 260 characters.
        //
        //   T:System.IO.DirectoryNotFoundException:
        //     The specified path is invalid (for example, it is on an unmapped drive).
        //
        //   T:System.IO.IOException:
        //     The directory specified by destinationDirectoryName already exists. -or- The
        //     name of an entry in the archive is System.String.Empty, contains only white space,
        //     or contains at least one invalid character. -or- Extracting an archive entry
        //     would create a file that is outside the directory specified by destinationDirectoryName.
        //     (For example, this might happen if the entry name contains parent directory accessors.)
        //     -or- An archive entry to extract has the same name as an entry that has already
        //     been extracted from the same archive.
        //
        //   T:System.UnauthorizedAccessException:
        //     The caller does not have the required permission to access the archive or the
        //     destination directory.
        //
        //   T:System.NotSupportedException:
        //     destinationDirectoryName or sourceArchiveFileName contains an invalid format.
        //
        //   T:System.IO.FileNotFoundException:
        //     sourceArchiveFileName was not found.
        //
        //   T:System.IO.InvalidDataException:
        //     The archive specified by sourceArchiveFileName is not a valid zip archive. -or-
        //     An archive entry was not found or was corrupt. -or- An archive entry was compressed
        //     by using a compression method that is not supported.
```

And we'd catch that exception on line 429:

```
                catch (Exception ex)
                {
                    localError = ex.Message;
                    installSuccessful = false;
                }
```

So I *think* we're OK here, but please double check my reading of the code.